File: chat-uikit/src/main/java/io/agora/chat/uikit/chat/viewholder/EaseChatViewHolderFactory.java
package io.agora.chat.uikit.chat.viewholder;

import android.view.ViewGroup;

import androidx.annotation.NonNull;

import io.agora.chat.ChatMessage;
import io.agora.chat.uikit.adapter.EaseBaseRecyclerViewAdapter;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryCombineViewHolder;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryFileViewHolder;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryImageViewHolder;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryTextViewHolder;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryVideoViewHolder;
import io.agora.chat.uikit.chathistory.viewholder.EaseHistoryVoiceViewHolder;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryCombine;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryFile;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryImage;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryText;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryVideo;
import io.agora.chat.uikit.chathistory.widget.EaseChatRowHistoryVoice;
import io.agora.chat.uikit.chatthread.widget.EaseChatRowThreadNotify;
import io.agora.chat.uikit.constants.EaseConstant;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowCombine;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowCustom;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowFile;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowImage;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowText;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowUnknown;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowUnsent;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowVideo;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowVoice;

public class EaseChatViewHolderFactory {
    public static EaseBaseRecyclerViewAdapter.ViewHolder<ChatMessage> createViewHolder(@NonNull ViewGroup parent, EaseMessageViewType viewType) {

        switch (viewType) {
            case VIEW_TYPE_MESSAGE_TXT_ME:
            case VIEW_TYPE_MESSAGE_TXT_OTHER:
                return new EaseTextViewHolder(new EaseChatRowText(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME));
            case VIEW_TYPE_MESSAGE_IMAGE_ME:
            case VIEW_TYPE_MESSAGE_IMAGE_OTHER:
                return new EaseImageViewHolder(new EaseChatRowImage(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME));
            case VIEW_TYPE_MESSAGE_VIDEO_ME:
            case VIEW_TYPE_MESSAGE_VIDEO_OTHER:
                return new EaseVideoViewHolder(new EaseChatRowVideo(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME));
            case VIEW_TYPE_MESSAGE_VOICE_ME:
            case VIEW_TYPE_MESSAGE_VOICE_OTHER:
                return new EaseVoiceViewHolder(new EaseChatRowVoice(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME));
            case VIEW_TYPE_MESSAGE_FILE_ME:
            case VIEW_TYPE_MESSAGE_FILE_OTHER:
                return new EaseFileViewHolder(new EaseChatRowFile(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME));
            case VIEW_TYPE_MESSAGE_CUSTOM_ME:
            case VIEW_TYPE_MESSAGE_CUSTOM_OTHER:
                return new EaseCustomViewHolder(new EaseChatRowCustom(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME));
            case VIEW_TYPE_MESSAGE_COMBINE_ME:
            case VIEW_TYPE_MESSAGE_COMBINE_OTHER:
                return new EaseCombineViewHolder(new EaseChatRowCombine(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME));
            case VIEW_TYPE_MESSAGE_UNSENT_ME:
            case VIEW_TYPE_MESSAGE_UNSENT_OTHER:
                return new EaseUnsentViewHolder(new EaseChatRowUnsent(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_ME));
            case VIEW_TYPE_MESSAGE_CHAT_THREAD_NOTIFY:
                return new EaseThreadNotifyViewHolder(new EaseChatRowThreadNotify(parent.getContext(), false));
            case VIEW_TYPE_MESSAGE_UNKNOWN_ME:
            case VIEW_TYPE_MESSAGE_UNKNOWN_OTHER:
                return new EaseUnknownViewHolder(new EaseChatRowUnknown(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME));
            default:
                return new EaseUnknownViewHolder(new EaseChatRowUnknown(parent.getContext(), false));
        }
    }

    public static EaseBaseRecyclerViewAdapter.ViewHolder<ChatMessage> createHistoryViewHolder(@NonNull ViewGroup parent, EaseMessageViewType viewType) {

        switch (viewType) {
            case VIEW_TYPE_MESSAGE_TXT_ME:
            case VIEW_TYPE_MESSAGE_TXT_OTHER:
                return new EaseHistoryTextViewHolder(new EaseChatRowHistoryText(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME));
            case VIEW_TYPE_MESSAGE_IMAGE_ME:
            case VIEW_TYPE_MESSAGE_IMAGE_OTHER:
                return new EaseHistoryImageViewHolder(new EaseChatRowHistoryImage(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME));
            case VIEW_TYPE_MESSAGE_VIDEO_ME:
            case VIEW_TYPE_MESSAGE_VIDEO_OTHER:
                return new EaseHistoryVideoViewHolder(new EaseChatRowHistoryVideo(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME));
            case VIEW_TYPE_MESSAGE_VOICE_ME:
            case VIEW_TYPE_MESSAGE_VOICE_OTHER:
                return new EaseHistoryVoiceViewHolder(new EaseChatRowHistoryVoice(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME));
            case VIEW_TYPE_MESSAGE_FILE_ME:
            case VIEW_TYPE_MESSAGE_FILE_OTHER:
                return new EaseHistoryFileViewHolder(new EaseChatRowHistoryFile(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME));
            case VIEW_TYPE_MESSAGE_CUSTOM_ME:
            case VIEW_TYPE_MESSAGE_CUSTOM_OTHER:
                return new EaseCustomViewHolder(new EaseChatRowCustom(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME));
            case VIEW_TYPE_MESSAGE_COMBINE_ME:
            case VIEW_TYPE_MESSAGE_COMBINE_OTHER:
                return new EaseHistoryCombineViewHolder(new EaseChatRowHistoryCombine(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME));
            case VIEW_TYPE_MESSAGE_UNKNOWN_ME:
            case VIEW_TYPE_MESSAGE_UNKNOWN_OTHER:
                return new EaseUnknownViewHolder(new EaseChatRowUnknown(parent.getContext(), viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME));
            default:
                return new EaseUnknownViewHolder(new EaseChatRowUnknown(parent.getContext(), false));
        }
    }

    public static int getViewType(@NonNull ChatMessage message) {
        return getChatType(message).getValue();
    }

    public static EaseMessageViewType getChatType(@NonNull ChatMessage message) {
        EaseMessageViewType type;
        ChatMessage.Type messageType = message.getType();
        ChatMessage.Direct direct = message.direct();
        if (messageType == ChatMessage.Type.TXT) {
            boolean isThreadNotify = message.getBooleanAttribute(EaseConstant.EASE_THREAD_NOTIFICATION_TYPE, false);
            boolean isRecallMessage = message.getBooleanAttribute(EaseConstant.MESSAGE_TYPE_RECALL, false);
            if(isThreadNotify) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_CHAT_THREAD_NOTIFY;
            }else if(isRecallMessage) {
                if(direct == ChatMessage.Direct.SEND) {
                    type = EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_ME;
                }else {
                    type = EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_OTHER;
                }
            }else {
                if(direct == ChatMessage.Direct.SEND) {
                    type = EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME;
                }else {
                    type = EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_OTHER;
                }
            }
        } else if (messageType == ChatMessage.Type.IMAGE) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_OTHER;
            }
        } else if (messageType == ChatMessage.Type.VIDEO) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_OTHER;
            }
        } else if (messageType == ChatMessage.Type.LOCATION) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_LOCATION_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_LOCATION_OTHER;
            }
        } else if (messageType == ChatMessage.Type.VOICE) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_OTHER;
            }
        } else if (messageType == ChatMessage.Type.FILE) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_OTHER;
            }
        } else if (messageType == ChatMessage.Type.CMD) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_CMD_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_CMD_OTHER;
            }
        } else if (messageType == ChatMessage.Type.CUSTOM) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_OTHER;
            }
        } else if (messageType == ChatMessage.Type.COMBINE) {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_OTHER;
            }
        } else {
            if(direct == ChatMessage.Direct.SEND) {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME;
            }else {
                type = EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_OTHER;
            }
        }
        return type;
    }
}

File: ease-im-kit/src/main/kotlin/com/hyphenate/easeui/feature/chat/viewholders/EaseChatViewHolderFactory.kt
package com.hyphenate.easeui.feature.chat.viewholders

import android.view.ViewGroup
import com.hyphenate.easeui.base.EaseBaseRecyclerViewAdapter
import com.hyphenate.easeui.common.ChatCustomMessageBody
import com.hyphenate.easeui.common.ChatMessage
import com.hyphenate.easeui.common.ChatMessageDirection
import com.hyphenate.easeui.common.ChatMessageType
import com.hyphenate.easeui.common.EaseConstant
import com.hyphenate.easeui.widget.chatrow.EaseChatRowCustom
import com.hyphenate.easeui.widget.chatrow.EaseChatRowFile
import com.hyphenate.easeui.widget.chatrow.EaseChatRowImage
import com.hyphenate.easeui.widget.chatrow.EaseChatRowText
import com.hyphenate.easeui.widget.chatrow.EaseChatRowUnknown
import com.hyphenate.easeui.widget.chatrow.EaseChatRowUnsent
import com.hyphenate.easeui.widget.chatrow.EaseChatRowUserCard
import com.hyphenate.easeui.widget.chatrow.EaseChatRowVideo
import com.hyphenate.easeui.widget.chatrow.EaseChatRowVoice
import com.hyphenate.easeui.widget.chatrow.EaseChatRowAlert
import com.hyphenate.easeui.widget.chatrow.EaseChatRowThreadNotify
import com.hyphenate.easeui.widget.chatrow.EaseChatRowCombine

object EaseChatViewHolderFactory {
    fun createViewHolder(
        parent: ViewGroup,
        viewType: EaseMessageViewType
    ): EaseBaseRecyclerViewAdapter.ViewHolder<ChatMessage> {
        return when (viewType) {
            EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_OTHER -> EaseTextViewHolder(
                EaseChatRowText(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_OTHER -> EaseImageViewHolder(
                EaseChatRowImage(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_OTHER -> EaseVideoViewHolder(
                EaseChatRowVideo(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_OTHER -> EaseVoiceViewHolder(
                EaseChatRowVoice(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_OTHER -> EaseFileViewHolder(
                EaseChatRowFile(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_OTHER -> EaseCustomViewHolder(
                EaseChatRowCustom(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_OTHER -> EaseCombineViewHolder(
                EaseChatRowCombine(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_OTHER -> EaseUnsentViewHolder(
                EaseChatRowUnsent(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_USER_CARD_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_USER_CARD_OTHER -> EaseUserCardViewHolder(
                EaseChatRowUserCard(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_USER_CARD_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_ALERT -> EaseAlertViewHolder(
                EaseChatRowAlert(parent.context)
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME, EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_OTHER -> EaseUnknownViewHolder(
                EaseChatRowUnknown(
                    parent.context,
                    isSender = viewType == EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME
                )
            )

            EaseMessageViewType.VIEW_TYPE_MESSAGE_CHAT_THREAD_NOTIFY -> EaseThreadNotifyViewHolder(
                EaseChatRowThreadNotify(
                    parent.context
                )
            )

            else -> EaseUnknownViewHolder(EaseChatRowUnknown(parent.context, isSender = false))
        }
    }

    fun getViewType(message: ChatMessage?): Int {
        return message?.let { getChatType(it).value } ?: EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_OTHER.value
    }

    fun getChatType(message: ChatMessage): EaseMessageViewType {
        val type: EaseMessageViewType
        val messageType = message.type
        val direct = message.direct()
        type = if (messageType == ChatMessageType.TXT) {
            val isThreadNotify: Boolean =
                message.getBooleanAttribute(EaseConstant.THREAD_NOTIFICATION_TYPE, false)
            val isRecallMessage: Boolean =
                message.getBooleanAttribute(EaseConstant.MESSAGE_TYPE_RECALL, false)
            val isContactNotify: Boolean =
                message.getBooleanAttribute(EaseConstant.MESSAGE_TYPE_CONTACT_NOTIFY, false)
            if (isThreadNotify) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_CHAT_THREAD_NOTIFY
            } else if (isRecallMessage || isContactNotify) {
                if (direct == ChatMessageDirection.SEND) {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_ME
                } else {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_UNSENT_OTHER
                }
            } else {
                if (direct == ChatMessageDirection.SEND) {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_ME
                } else {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_TXT_OTHER
                }
            }
        } else if (messageType == ChatMessageType.IMAGE) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_IMAGE_OTHER
            }
        } else if (messageType == ChatMessageType.VIDEO) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_VIDEO_OTHER
            }
        } else if (messageType == ChatMessageType.LOCATION) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_LOCATION_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_LOCATION_OTHER
            }
        } else if (messageType == ChatMessageType.VOICE) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_VOICE_OTHER
            }
        } else if (messageType == ChatMessageType.FILE) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_FILE_OTHER
            }
        } else if (messageType == ChatMessageType.CMD) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_CMD_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_CMD_OTHER
            }
        } else if (messageType == ChatMessageType.CUSTOM) {
            val event = (message.body as? ChatCustomMessageBody)?.event() ?: ""
            if (event == EaseConstant.USER_CARD_EVENT) {
                if (direct == ChatMessageDirection.SEND) {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_USER_CARD_ME
                } else {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_USER_CARD_OTHER
                }
            } else if (event == EaseConstant.MESSAGE_CUSTOM_ALERT) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_ALERT
            } else {
                if (direct == ChatMessageDirection.SEND) {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_ME
                } else {
                    EaseMessageViewType.VIEW_TYPE_MESSAGE_CUSTOM_OTHER
                }
            }
        } else if (messageType == ChatMessageType.COMBINE) {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_COMBINE_OTHER
            }
        } else {
            if (direct == ChatMessageDirection.SEND) {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_ME
            } else {
                EaseMessageViewType.VIEW_TYPE_MESSAGE_UNKNOWN_OTHER
            }
        }
        return type
    }
}